### PR TITLE
Victory crash/freeze fix

### PR DESCRIPTION
**Pull request by dydzio0614**
This PR fixes bugs that cause crashes/freezes on scenario/campaign map win. Default handling for lost connection changed to avoid crash (do not try to close server). Changing server state to GAMEPLAY_ENDED is properly handled, it left server in unstable state before, causing unavoidable assert failure. And last, thread synchronization has been made between closing server and restarting it for next campaign mission. I guess it can be implemented to work for every server close/start case in the code, so client cannot join exiting server, but I did not dig into it.

Tl;dr: Map finish problems fixed, campaigns work again.

### CODE REVIEW

**Comment by nullkiller** [comment]
There is visit town problem at least for me. When game ends by capturing last AI town. It looks different but I quite often used to end campaign games such way.

**Comment by ArseniyShestakov** [comment]
There are also at least two crashes in MP:
- When one of clients wins, everyone else crashes.
- Seems like server crashes currently when just one player lose or might be when he's disconnected...
Yet I can merge this one now since there certainly no regressions.
